src/stm32f1/timer1_pwm.cpp

```diff
@@ -0,0 +1,203 @@
+      .insert<output_compare_even>(pwm_mode_1)
+      .set(even_channel_preload_enable);
+    setup_channel(reg, 4);
+  }
+
+  bit_modify(reg->control_register).set(counter_enable);
+
+  reg->prescale_register = 0x0U;
+
+  // Set counter value to 0 because it upcounts.
+  reg->counter_register = 0x0U;
+
+  reg->auto_reload_register = 0xFFFF;
+}
+
+}  // namespace
+
+timer1_pwm::timer1_pwm(timer1_pwm::pwm_pins p_pin)
+  : m_pin{ p_pin }
+{
+  // config pin to alternate function push - pull
+  auto const p_id = get_peripheral_id(p_pin);
+  auto const p_reg = get_pwm_reg(p_id);
+  power_on(p_id);
+  // power_on(peripheral::afio);
+  switch (p_pin) {
+    case timer1_pwm::pwm_pins::pa10:
+      configure_pin({ .port = 'A', .pin = 10 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_3;
+      m_channel = 3;
+      break;
+    case timer1_pwm::pwm_pins::pa11:
+      configure_pin({ .port = 'A', .pin = 11 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_4;
+      m_channel = 4;
+      break;
+    case timer1_pwm::pwm_pins::pa8:
+      configure_pin({ .port = 'A', .pin = 8 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register;
+      m_channel = 1;
+      break;
+    case timer1_pwm::pwm_pins::pa9:
+      configure_pin({ .port = 'A', .pin = 9 }, push_pull_alternative_output);
+      m_compare_register_addr = &p_reg->capture_compare_register_2;
+      m_channel = 2;
+      break;
+    default:
+      safe_throw(hal::operation_not_supported(this));
+  }
+  setup(p_pin);
+}
+void timer1_pwm::driver_frequency(hertz p_frequency)
+{
+  auto const peripheral_id = get_peripheral_id(m_pin);
+  pwm_reg_t* reg = get_pwm_reg(peripheral_id);
+
+  auto const current_clock = hal::stm32f1::frequency(peripheral_id);
+
+  auto const previous_duty_cycle =
+    get_duty_cycle(m_pin, m_compare_register_addr);
+
+  if (p_frequency >= current_clock) {
+    safe_throw(hal::operation_not_supported(this));
+  }
+
+  auto const possible_prescaler_value =
+    static_cast<float>(current_clock / (p_frequency * std::pow(2, 16)));
+
+  std::uint16_t prescale = 0;
+  std::uint16_t autoreload = 0xFFFF;
+
+  if (possible_prescaler_value > 1) {
+    prescale = std::floor(possible_prescaler_value);
+  } else {
+    // we have to reduce the ARR value.
+    autoreload = static_cast<std::uint16_t>(current_clock / p_frequency);
+  }
+  reg->auto_reload_register = autoreload;
+  reg->prescale_register = prescale;
+
+  // Update duty cycle according to new frequency
+  driver_duty_cycle(previous_duty_cycle);
+}
+void timer1_pwm::driver_duty_cycle(float p_duty_cycle)
+{
+  auto const peripheral_id = get_peripheral_id(m_pin);
+  pwm_reg_t* reg = get_pwm_reg(peripheral_id);
+
+  std::uint16_t desired_ccr_value =
+    (static_cast<std::uint16_t>(reg->auto_reload_register) *
+     static_cast<std::uint16_t>(p_duty_cycle));
```

Comment:
```suggestion
     static_cast<std::uint16_t>(p_duty_cycle));
```

The `p_duty_cycle` is a float between 0 to 1. cast it to a u16 would basically result in the value being 0 or 1. We shouldn't cast the float before we used it as a float. You can cast the resultant to the desired_ccr_value.